Sources/Segment/Errors.swift

```diff
@@ -0,0 +1,79 @@
+//
+//  Errors.swift
+//  
+//
+//  Created by Brandon Sneed on 10/20/22.
+//
+
+import Foundation
+
+public enum AnalyticsError: Error {
+    case storageUnableToCreate(String)
+    case storageUnableToWrite(String)
+    case storageUnableToRename(String)
+    case storageUnableToOpen(String)
+    case storageInvalid(String)
+    case storageUnknown(Error)
+
+    case networkUnexpectedHTTPCode(Int)
+    case networkServerLimited(Int)
+    case networkServerRejected(Int)
+    case networkUnknown(Error)
+    case networkInvalidData
+
+    case jsonUnableToSerialize(Error)
+    case jsonUnableToDeserialize(Error)
+    case jsonUnknown(Error)
+
+    case pluginError(Error)
+}
+
+extension Analytics {
+    /// Tries to convert known error types to AnalyticsError.
+    static internal func translate(error: Error) -> Error {
+        if let e = error as? OutputFileStream.OutputStreamError {
+            switch e {
+            case .invalidPath(let path):
+                return AnalyticsError.storageInvalid(path)
+            case .unableToCreate(let path):
+                return AnalyticsError.storageUnableToCreate(path)
+            case .unableToOpen(let path):
+                return AnalyticsError.storageUnableToOpen(path)
+            case .unableToWrite(let path):
+                return AnalyticsError.storageUnableToWrite(path)
+            }
+        }
+        
+        if let e = error as? JSON.JSONError {
+            switch e {
+            case .incorrectType:
+                return AnalyticsError.jsonUnableToDeserialize(e)
+            case .nonJSONType:
+                return AnalyticsError.jsonUnableToDeserialize(e)
+            case .unknown:
+                return AnalyticsError.jsonUnknown(e)
+            }
+        }
+        return error
+    }
+    
+    /// Reports an internal error to the user-defined error handler.
+    public func reportInternalError(_ error: Error, hardStop: Bool = false) {
```

Comment:
nit: maybe `critical` instead of `hardStop` sounds more exception-like?

Reply:
Good call.

Reply:
named it `fatal` instead as it seemed more descriptive of what was going to happen, ie: YOUR APP WILL DIE.